### PR TITLE
Warn state for CPU, Memory, Filesystem and Battery modules

### PR DESCRIPTION
**Pull request by DRKblade**
This is an [item](https://github.com/polybar/polybar/projects/7#card-25459423) in the todo list from the [Module State Machines project](https://github.com/polybar/polybar/projects/7), and was requested in #570 and #956. The idea is that the modules that monitor system resources will display different labels and animations when that resource requires attention (battery low, high temperature, high cpu/memory usage).

The warn state would work like that of the temperature module. The condition to activate the warn states are as follow:
* Memory module: percentage of memory used >= `warn-percentage`
* CPU module: total percentage of all cores >= `warn-percentage`
* Filesystem module: the filesystem is mounted and percentage of used space >= `warn-percentage`
* Battery module: the charge percentage >= `low-at`

**Release Notes:**

This adds `format-warn` and `label-warn` to the memory, cpu, fs, and battery module.

It also slightly changes how the ramps in those modules behave. The ramps are bounded by these warn thresholds and, the same as the temperature module, the first and last ramp elements are responsible for everything `<=` the lower bound or `>=` the upper bound. Before, the ramp generally went from 0 to 100 and the first and last ramp elements covered the same share inside the range as all other ramp elements as well as everything outside the range.

These two changes in behavior partially cancel each other out, so the change in behavior in existing configs will be barely noticeable.

It also adds the following:

* battery: `animation-low`, `low-at = 10`
* cpu: `warn-percentage = 80`
* fs: `warn-percentage = 90`
* memory: `warn-percentage = 90`

**/Release Notes**

Example configuration:
```
[module/cpu]
type = internal/cpu

interval = 1
warn-percentage = 80

format = <label>
format-prefix = CPU
format-background = #000
format-foreground = #fff
format-padding = 1
label = " %percentage%%"

format-warn = <label-warn>
format-warn-prefix = CPU
format-warn-background = #fff
format-warn-foreground = #000
format-warn-padding = 2
label-warn = " %percentage%%"
```

### CODE REVIEW

**Comment by DRKblade** [comment]
This would also provide the feature requested by #1871.

File: src/modules/cpu.cpp [hunk 1]
```diff
@@ -17,15 +17,22 @@ namespace modules {
 
   cpu_module::cpu_module(const bar_settings& bar, string name_) : timer_module<cpu_module>(bar, move(name_)) {
     m_interval = m_conf.get<decltype(m_interval)>(name(), "interval", 1s);
-
+    m_totalwarn = m_conf.get(name(), "warn-percentage", 80);
```
Comment (patrick96) [inline src/modules/cpu.cpp:20]:
```suggestion
    m_totalwarn = m_conf.get(name(), "warn-percentage", m_totalwarn);
```

File: src/modules/fs.cpp [hunk 2]
```diff
@@ -104,6 +110,31 @@ namespace modules {
...
+	if (m_labelmounted) {
+	  replace_tokens(m_labelmounted);
+	}
+	if (m_labelwarn) {
+	  replace_tokens(m_labelwarn);
+	}
+	if (m_labelunmounted) {
+          m_labelunmounted->reset_tokens();
+          m_labelunmounted->replace_token("%mountpoint%", mount->mountpoint);
+	}
```
Comment (patrick96) [inline src/modules/fs.cpp:137]: You have some weird indentations here.

File: tests/unit_tests/drawtypes/ramp.cpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  r.add(factory_util::shared<label>("test1", 0));
+  r.add(factory_util::shared<label>("test2", 0));
+  r.add(factory_util::shared<label>("test3", 0));
+  EXPECT_EQ("test1", r.get_by_percentage(33)->get());
+  EXPECT_EQ("test2", r.get_by_percentage(34)->get());
+  EXPECT_EQ("test3", r.get_by_percentage(67)->get());
+  EXPECT_EQ("test1", r.get_by_percentage_with_borders(19, 20, 40)->get());
+  EXPECT_EQ("test2", r.get_by_percentage_with_borders(21, 20, 40)->get());
+  EXPECT_EQ("test2", r.get_by_percentage_with_borders(39, 20, 40)->get());
+  EXPECT_EQ("test3", r.get_by_percentage_with_borders(41, 20, 40)->get());
```
Comment (patrick96) [inline tests/unit_tests/drawtypes/ramp.cpp:19]: Please also add the following tests:

```cpp
  EXPECT_EQ("test1", r.get_by_percentage_with_borders(20, 20, 40)->get());
  EXPECT_EQ("test3", r.get_by_percentage_with_borders(40, 20, 40)->get());
```

We should test these edge cases as well.

**Review by patrick96 — CHANGES_REQUESTED** [review]
Code looks good. Thanks!

I do have some concerns about how this changes existing behavior. Because the thresholds for the new warn states would change how existing configs would display once we are in the warn state.

For example right now, if your CPU is over 80%, it would display whatever is in `format` and the labels and ramps used there. If we used the same config in this PR, it would now display just the total load percentage.

We generally want people to be able to update without it changing how their bar looks like (at least not in a substantial way). So we can't really have these warning states enabled by default, or we must make sure that they behave exactly the same as the non-warn states by default.

I can think of two solutions to this:

1. Set the threshold to something that can never be reached. In the battery module, we would set it to anything smaller than `0`. For memory, cpu, and fs, we would have to set it to something larger than 100. It has to be strictly smaller than 0 or larger than 100 because both 0% battery and 100% cpu, ram, or storage can be reached and the warn states are activated *at* those values.
2. Have an extra boolean flag in each module that is set to true only if the new warn formats were defined in the module config. Only use the warn states if the flag is true.

None of these solutions are pretty unfortunately. Can you think of another way of ensuring that the module doesn't suddenly display something completely different in the warn states?

---

And just to note, because it also changes module behavior: This PR adds the same changes to the behavior of the first and last ramp elements for the battery, fs, cpu, and memory module as #2197 added for the temperature module.

While this can be considered a breaking change, it just changes the ranges that ramp elements are responsible for a bit. So, the same as for the temperature module, this is probably fine, we just need to make sure to mention it in the changelogs.

File: src/modules/fs.cpp (hunk 2 above)
Comment (DRKblade) [reply src/modules/fs.cpp:137]: I tried to fix this, but my editor may insert tabs instead of spaces. Tell me if it's still a problem after my commit

**Comment by DRKblade** [comment]
We can create a new fallback tag. The default value for FORMAT_WARN will be this tag, and when `module::get_format()` is called, we check if FORMAT_WARN have this tag to return DEFAULT_FORMAT instead.

**Comment by DRKblade** [comment]
Another idea on my mind is that I can write `module_formatter::try_get()` which will add a format only if it is defined. If we load the warn format using `try_get`, I can then write `module_formatter::has_format()` to check if the warn format exist and use the default format when it isn't.

**Comment by patrick96** [comment]
> Another idea on my mind is that I can write module_formatter::try_get() which will add a format only if it is defined. If we load the warn format using try_get, I can then write module_formatter::has_format() to check if the warn format exist and use the default format when it isn't.

I think that is a good idea. :+1:

**Comment by DRKblade** [comment]
I applied it on all relevant modules.

File: src/modules/fs.cpp [hunk 4]
```diff
@@ -141,7 +172,11 @@ namespace modules {
    * Select format based on fs state
    */
   string fs_module::get_format() const {
-    return m_mounts[m_index]->mounted ? FORMAT_MOUNTED : FORMAT_UNMOUNTED;
+    if (!m_mounts[m_index]->mounted)
+      return FORMAT_UNMOUNTED;
```
Comment (patrick96) [inline src/modules/fs.cpp:176]: Please add curly braces around all if statements.

File: src/modules/meta/base.cpp [hunk 5]
```diff
@@ -138,6 +137,15 @@ namespace modules {
 
     m_formats.insert(make_pair(move(name), move(format)));
   }
+  	
+  void module_formatter::add(string name, string fallback, vector<string>&& tags, vector<string>&& whitelist) {
+    add_value(move(name), m_conf.get(m_modname, move(name), move(fallback)), forward<vector<string>>(tags), forward<vector<string>>(whitelist));
+  }
+  	
+  void module_formatter::add(string name, vector<string>&& tags, vector<string>&& whitelist) {
+   	if (m_conf.has(m_modname, name))
```
Comment (patrick96) [inline src/modules/meta/base.cpp:146]: Please also add curly braces here. And you seem to have tab indentation here.

File: src/modules/meta/base.cpp [hunk 6]
```diff
@@ -138,6 +137,15 @@ namespace modules {
 
     m_formats.insert(make_pair(move(name), move(format)));
   }
+  	
+  void module_formatter::add(string name, string fallback, vector<string>&& tags, vector<string>&& whitelist) {
+    add_value(move(name), m_conf.get(m_modname, move(name), move(fallback)), forward<vector<string>>(tags), forward<vector<string>>(whitelist));
+  }
+  	
+  void module_formatter::add(string name, vector<string>&& tags, vector<string>&& whitelist) {
```
Comment (patrick96) [inline src/modules/meta/base.cpp:145]: We should probably name this `add_optional` to make it explicit what it is for.

**Review by patrick96 — CHANGES_REQUESTED** [review]
Looks good. Just some minor style changes.

Please also rebase and resolve the merge conflicts.

**Comment by DRKblade** [comment]
Ok, is it good?

File: src/modules/fs.cpp [hunk 7]
```diff
@@ -104,6 +110,31 @@ namespace modules {
...
+	if (m_labelmounted) {
+	  replace_tokens(m_labelmounted);
+	}
+	if (m_labelwarn) {
+	  replace_tokens(m_labelwarn);
+	}
+	if (m_labelunmounted) {
+    m_labelunmounted->reset_tokens();
+    m_labelunmounted->replace_token("%mountpoint%", mount->mountpoint);
+	}
```
Comment (patrick96) [inline src/modules/fs.cpp:137]: Please indent with spaces

File: src/modules/meta/base.cpp [hunk 8]
```diff
@@ -140,6 +138,16 @@ namespace modules {
...
   }
+  	
+  void module_formatter::add(string name, string fallback, vector<string>&& tags, vector<string>&& whitelist) {
+    add_value(move(name), m_conf.get(m_modname, move(name), move(fallback)), forward<vector<string>>(tags), forward<vector<string>>(whitelist));
+  }
+  	
+  void module_formatter::add_optional(string name, vector<string>&& tags, vector<string>&& whitelist) {
+   	if (m_conf.has(m_modname, name)) {
+      add_value(move(name), m_conf.get(m_modname, move(name)), move(tags), move(whitelist));
+   	}
```
Comment (patrick96) [inline src/modules/meta/base.cpp:149]: Please reindent with spaces

File: src/modules/meta/base.cpp [hunk 9]
```diff
@@ -86,13 +86,12 @@ namespace modules {
   // }}}
   // module_formatter {{{
 
-  void module_formatter::add(string name, string fallback, vector<string>&& tags, vector<string>&& whitelist) {
-    const auto formatdef = [&](const string& param, const auto& fallback) {
-      return m_conf.get("settings", "format-" + param, fallback);
-    };
+	void module_formatter::add_value(string&& name, string&& value, vector<string>&& tags, vector<string>&& whitelist) {
```
Comment (patrick96) [inline src/modules/meta/base.cpp:89]: Please reindent with spaces

File: src/modules/fs.cpp [hunk 10]
```diff
@@ -141,7 +172,12 @@ namespace modules {
    * Select format based on fs state
    */
   string fs_module::get_format() const {
-    return m_mounts[m_index]->mounted ? FORMAT_MOUNTED : FORMAT_UNMOUNTED;
+    if (!m_mounts[m_index]->mounted)
+      return FORMAT_UNMOUNTED;
+    if (m_mounts[m_index]->percentage_used >= m_perc_used_warn && m_formatter->has(FORMAT_WARN)) {
```
Comment (patrick96) [inline src/modules/fs.cpp:177]:
```suggestion
    if (m_mounts[m_index]->percentage_used >= m_perc_used_warn && m_formatter->has_format(FORMAT_WARN)) {
```

**Review by patrick96 — CHANGES_REQUESTED** [review]
Functionality wise, everything works now. There was a small bug in the fs module, but I added a comment for that.

The build currently breaks, I have added comments for all the places that need to be fixed.

There are also some places where you use tabs instead of spaces, I have also added comments for that.

After these changes, I think we can merge this.

File: src/modules/fs.cpp [hunk 11]
```diff
@@ -28,18 +28,24 @@ namespace modules {
...
     m_spacing = m_conf.get(name(), "spacing", m_spacing);
     m_interval = m_conf.get<decltype(m_interval)>(name(), "interval", 30s);
 
     // Add formats and elements
     m_formatter->add(
         FORMAT_MOUNTED, TAG_LABEL_MOUNTED, {TAG_LABEL_MOUNTED, TAG_BAR_FREE, TAG_BAR_USED, TAG_RAMP_CAPACITY});
-    m_formatter->add(FORMAT_UNMOUNTED, TAG_LABEL_UNMOUNTED, {TAG_LABEL_UNMOUNTED});
+    m_formatter->add(
+        FORMAT_WARN, {TAG_LABEL_WARN, TAG_BAR_FREE, TAG_BAR_USED, TAG_RAMP_CAPACITY});
+    m_formatter->add_optional(FORMAT_UNMOUNTED, TAG_LABEL_UNMOUNTED, {TAG_LABEL_UNMOUNTED});
```
[the remark this answers is not in the document]
Comment (DRKblade) [reply src/modules/fs.cpp:41]: Oops, I thought I have committed that change before pushing.

**Comment by DRKblade** [comment]
I think I finally got the indents right.

**Review by patrick96 — APPROVED** [review]
Everything looks good now :tada:

Thanks a lot for all your efforts! :smiley:
